### PR TITLE
Add mcxn236 basic environment

### PR DESCRIPTION
**Pull request by NeilChen93**
Enable mcxn236 basic environment：
Test below cases: hello_world, button, blinky, synchronization, philosophers

### CODE REVIEW

File: boards/nxp/frdm_mcxn236/Kconfig.defconfig
Comment (danieldegrasse) [inline boards/nxp/frdm_mcxn236/Kconfig.defconfig:1]: If no changes to Kconfig defaults are needed for this board, we can remove this file

File: boards/nxp/frdm_mcxn236/board.c [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	/* Enable SCG clock */
+	CLOCK_EnableClock(kCLOCK_Scg);
+
+	/* FRO OSC setup - begin, enable the FRO for safety switching */
+
+	/* Switch to FRO 12M first to ensure we can change the clock setting */
+	CLOCK_AttachClk(kFRO12M_to_MAIN_CLK);
+
+	/* Configure Flash wait-states to support 1.2V voltage level and 150000000Hz frequency */
+	FMU0->FCTRL = (FMU0->FCTRL & ~((uint32_t)FMU_FCTRL_RWSC_MASK)) | (FMU_FCTRL_RWSC(0x3U));
```
Comment (danieldegrasse) [inline boards/nxp/frdm_mcxn236/board.c:66]: Any reason not to use `CLOCK_SetFLASHAccessCyclesForFreq` here?

File: boards/nxp/frdm_mcxn236/board.cmake [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#
+# Copyright 2024 NXP
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+if(CONFIG_SOC_MCXN236)
```
Comment (danieldegrasse) [inline boards/nxp/frdm_mcxn236/board.cmake:7]: Why make this dependent on the SOC Kconfig? I expect it will always be set for this board, so we can probably drop this right?

File: boards/nxp/frdm_mcxn236/doc/index.rst [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+For both options use the ``-r jlink`` option with west to use the jlink runner.
+
+.. code-block:: console
+
+   west flash -r jlink
+
+Configuring a Console
+=====================
+
+Connect a USB cable from your PC to J17, and use the serial terminal of your choice
```
Comment (danieldegrasse) [inline boards/nxp/frdm_mcxn236/doc/index.rst:148]:
```suggestion
Connect a USB cable from your PC to J10, and use the serial terminal of your choice
```

It seems like this is the USB connector used for the debugger, right?

image

File: boards/nxp/frdm_mcxn236/frdm_mcxn236.dts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+&edma0 {
+	status = "okay";
+};
+
+&flexcomm2 {
+	status = "okay";
+};
+
+/*
+ *LPFLEXCOMM supports UART and I2C on the same instance, enable this for
```
Comment (danieldegrasse) [inline boards/nxp/frdm_mcxn236/frdm_mcxn236.dts:69]:
```suggestion
 * LPFLEXCOMM supports UART and I2C on the same instance, enable this for
```

File: boards/nxp/frdm_mcxn236/frdm_mcxn236.dtsi [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+		 */
+		slot0_partition: partition@10000 {
+			label = "image-0";
+			reg = <0x00010000 DT_SIZE_K(480)>;
+		};
+		slot1_partition: partition@80000 {
+			label = "image-1";
+			reg = <0x0088000 DT_SIZE_K(472)>;
+		};
+		/* storage_partition is placed in WINBOND flash memory*/
```
Comment (danieldegrasse) [inline boards/nxp/frdm_mcxn236/frdm_mcxn236.dtsi:104]: It doesn't seem like we have the winbond flash memory defined for this board yet in the devicetree, so we should probably drop this comment for the initial board enablement

File: boards/nxp/frdm_mcxn236/frdm_mcxn236.yaml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+# Copyright 2024 NXP
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+identifier: frdm_mcxn236
+name: NXP FRDM MCXN236
+type: mcu
+arch: arm
+ram: 192
```
Comment (danieldegrasse) [inline boards/nxp/frdm_mcxn236/frdm_mcxn236.yaml:11]: It looks like the `sram0` bank we are using for RAM is 256KB, right? So we should make this value 256 as well

File: dts/arm/nxp/nxp_mcxn23x_common.dtsi [hunk 7]
```diff
@@ -0,0 +1,692 @@
...
+		status = "disabled";
+
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		flash: flash@0 {
+			compatible = "soc-nv-flash";
+			reg = <0 DT_SIZE_M(1)>;
+			erase-block-size = <8192>;
+			write-block-size = <16>;
```
Comment (danieldegrasse) [inline dts/arm/nxp/nxp_mcxn23x_common.dtsi:502]: I know for the MCXN94x that the FMU itself supports 16 byte writes, but the ROM API the flash driver uses only supports 128 byte writes. Is that the case for this SOC as well?

File: tests/drivers/spi/spi_loopback/boards/frdm_mcxn236.overlay
Comment (danieldegrasse) [inline tests/drivers/spi/spi_loopback/boards/frdm_mcxn236.overlay:1]: Please split the changes to test SPI on this board into a new commit from the one adding I2C test support

File: tests/drivers/spi/spi_loopback/boards/frdm_mcxn236.overlay [hunk 8]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&flexcomm3_lpspi3 {
+	slow@0 {
```
Comment (danieldegrasse) [inline tests/drivers/spi/spi_loopback/boards/frdm_mcxn236.overlay:8]: Could you document the pins that should be shorted together in this overlay file, similar to what is done for the I2C overlay?

**Review by danieldegrasse — COMMENTED** [review]
Commits in this PR should be ordered so that each is bisectable. This way so that when a feature is added in a commit (such as SOC or board support), it will be possible for a developer to build for that board or SOC at the point of that commit in the repo history.

I think the commits in this PR should be ordered like so (oldest commit to newest)
- update HAL NXP manifest
- add dts for MCXN23x
- add SOC support for MCXN23x
- remove `fsl_power.h` for MCXN236
- add frdm_mcxn236 board
- add i2c test support for frdm_mcxn236 board
- add spi test support for frdm_mcxn236 board

File: dts/arm/nxp/nxp_mcxn23x_common.dtsi (hunk 7 above)
Comment (NeilChen93) [reply dts/arm/nxp/nxp_mcxn23x_common.dtsi:502]: About FMU, MCXN23x should be same as MCXN94x.

Comment (butok) [reply dts/arm/nxp/nxp_mcxn23x_common.dtsi:502]: > About FMU, MCXN23x should be same as MCXN94x.

According to the MCXN23xRM documentation, the Flash ROM API supports only 128 Bytes. So the write-block-size must be set to 128, as for MCXN94x.

File: boards/nxp/frdm_mcxn236/board.c (hunk 1 above)
Comment (NeilChen93) [reply boards/nxp/frdm_mcxn236/board.c:66]: This was copied from frdm_mcxn947 board.c.

File: tests/drivers/spi/spi_loopback/boards/frdm_mcxn236.overlay (hunk 8 above)
Comment (NeilChen93) [reply tests/drivers/spi/spi_loopback/boards/frdm_mcxn236.overlay:11]: Updated.

File: tests/drivers/spi/spi_loopback/boards/frdm_mcxn236.overlay
Comment (NeilChen93) [reply tests/drivers/spi/spi_loopback/boards/frdm_mcxn236.overlay:1]: Updated.

File: boards/nxp/frdm_mcxn236/frdm_mcxn236.yaml (hunk 6 above)
Comment (NeilChen93) [reply boards/nxp/frdm_mcxn236/frdm_mcxn236.yaml]: Updated.

File: boards/nxp/frdm_mcxn236/frdm_mcxn236.dts (hunk 4 above)
Comment (NeilChen93) [reply boards/nxp/frdm_mcxn236/frdm_mcxn236.dts]: Updated.

File: boards/nxp/frdm_mcxn236/doc/index.rst (hunk 3 above)
Comment (NeilChen93) [reply boards/nxp/frdm_mcxn236/doc/index.rst]: Updated.

File: boards/nxp/frdm_mcxn236/board.cmake (hunk 2 above)
Comment (NeilChen93) [reply boards/nxp/frdm_mcxn236/board.cmake]: Updated.

File: dts/arm/nxp/nxp_mcxn23x_common.dtsi
Comment (danieldegrasse) [inline dts/arm/nxp/nxp_mcxn23x_common.dtsi:1]: This change to fix the flash write block size does not need to be a separate commit-it can be squashed into the commit introducing the devicetree for the MCXN23X, since that is also part of this PR.

**Review by danieldegrasse — COMMENTED** [review]
Thanks for the changes, PR is looking better:)- Could you rebase to fix the merge conflict and take a look at the compliance failures when you get a chance?

**Comment by dleach02** [comment]
@NeilChen93 I rebased to address the merge-conflict for you.

File: west.yml [hunk 9]
```diff
@@ -198,7 +198,7 @@ manifest:
       groups:
         - hal
     - name: hal_nxp
-      revision: 88a7d079eed470ecc94010784931823a85ac3b39
+      revision: pull/391/head
```
Comment (decsny) [inline west.yml:201]: you can put the manifest change in the commit where the changes are first needed

File: dts/arm/nxp/nxp_mcxn23x.dtsi
Comment (decsny) [inline dts/arm/nxp/nxp_mcxn23x.dtsi:1]: can the mcxn23x share anything with the mcxn94x? and have then a more common dtsi file?

File: dts/arm/nxp/nxp_mcxn23x_common.dtsi [hunk 10]
```diff
@@ -0,0 +1,692 @@
...
+			#size-cells = <1>;
+
+			mpu: mpu@e000ed90 {
+				compatible = "arm,armv8m-mpu";
+				reg = <0xe000ed90 0x40>;
+			};
+		};
+	};
+
+	/* Dummy pinctrl node, filled with pin mux options at board level */
```
Comment (decsny) [inline dts/arm/nxp/nxp_mcxn23x_common.dtsi:31]: remove this comment, this is how literally all the nodes work

**Comment by danieldegrasse** [comment]
Please take a look at the build failure on `tests/drivers/spi/spi_loopback/drivers.spi.loopback` for the MCXN236 board when you get a chance

File: soc/nxp/mcx/mcxn23x/pinctrl_soc.h
Comment (danieldegrasse) [inline soc/nxp/mcx/mcxn23x/pinctrl_soc.h:1]: The commit formatting the pin control file can be squashed into the one that adds the file.

As a general rule in Zephyr (even during the review process) we try to have a clean commit history- so a PR should not contain "fixup" commits like this one in response to reviews. Instead, when you make a change in response to review squash that change back into the commit that initially touched the file (or if the file is touched multiple times in a PR, the commit where you made the change you are now reworking)

File: soc/nxp/mcx/mcxn23x/pinctrl_soc.h [hunk 11]
```diff
@@ -16,23 +16,22 @@ extern "C" {
...
 
-#define Z_PINCTRL_MCX_PINCFG(node_id)					\
-	(PORT_PCR_DSE(DT_ENUM_IDX(node_id, drive_strength)) |			\
-	PORT_PCR_PS(DT_PROP(node_id, bias_pull_up)) |				\
-	PORT_PCR_PE(DT_PROP(node_id, bias_pull_up)) |				\
-	PORT_PCR_PE(DT_PROP(node_id, bias_pull_down)) |				\
-	PORT_PCR_ODE(DT_PROP(node_id, drive_open_drain)) |			\
-	PORT_PCR_SRE(DT_ENUM_IDX(node_id, slew_rate)) |				\
-	PORT_PCR_IBE(DT_PROP(node_id, input_enable)) |				\
+#define Z_PINCTRL_MCX_PINCFG(node_id)                     \
```
Comment (danieldegrasse) [inline soc/nxp/mcx/mcxn23x/pinctrl_soc.h:19]: Indentation still looks wrong here in github UI- maybe check your editor settings? Usually the easiest way to get these right is to indent the `\` characters with spaces instead of tabs (which CI won't complain about, if I remember right)

**Comment by hakehuang** [comment]
need rebase
<html>
<body>

CASE_NAME | ZEPHYR_VERSION | RESULT | PLATFORM | TEST_PATH
-- | -- | -- | -- | --
kernel.poll | v3.6.0-5877-g1b719f53dcc0 | Build failure | frdm_mcxn236 | tests/kernel/poll
kernel.poll.minimallibc | v3.6.0-5877-g1b719f53dcc0 | Build failure | frdm_mcxn236 | tests/kernel/poll

</body>
</html>

**Comment by danieldegrasse** [comment]
> need rebase
> CASE_NAME 	ZEPHYR_VERSION 	RESULT 	PLATFORM 	TEST_PATH
> kernel.poll 	v3.6.0-5877-g1b719f53dcc0 	Build failure 	frdm_mcxn236 	tests/kernel/poll
> kernel.poll.minimallibc 	v3.6.0-5877-g1b719f53dcc0 	Build failure 	frdm_mcxn236 	tests/kernel/poll

These have been fixed upstream- you can also just restart failing twister tests, as twister will rebase the PR onto `main` before it runs:)

**Review by danieldegrasse — DISMISSED** [review]

File: drivers/timer/mcux_os_timer.c [hunk 12]
```diff
@@ -16,7 +16,9 @@
 #include <zephyr/drivers/counter.h>
 #include <zephyr/pm/pm.h>
 #include "fsl_ostimer.h"
+#ifndef CONFIG_SOC_SERIES_MCXN23X
```
Comment (mmahadevan108) [inline drivers/timer/mcux_os_timer.c:19]: You could instead use this
`#if (DT_INST_PROP(0, wakeup_source))`

Comment (dleach02) [reply drivers/timer/mcux_os_timer.c:19]: @mmahadevan108 is this a "could" or "should"? Not clear to me why you would use this over the other?

File: west.yml (hunk 9 above)
Comment (dleach02) [reply west.yml:201]: @decsny is there a problem with it being first? 

Which commit would you want this part of? The SOC one?

File: dts/arm/nxp/nxp_mcxn23x_common.dtsi [hunk 13]
```diff
@@ -0,0 +1,691 @@
...
+		clocks = <&syscon MCUX_CTIMER4_CLK>;
+		mode = <0>;
+		input = <0>;
+		prescale = <0>;
+	};
+};
+
+&systick {
+	/*
+	 * MCXN94X relies by default on the OS Timer for system clock
```
[the remark this answers is not in the document]
Comment (NeilChen93) [reply dts/arm/nxp/nxp_mcxn23x_common.dtsi]: Updated

**Review by danieldegrasse — DISMISSED** [review]

**Review by dleach02 — DISMISSED** [review]

File: drivers/timer/mcux_os_timer.c (hunk 12 above)
Comment (mmahadevan108) [reply drivers/timer/mcux_os_timer.c:19]: Just to avoid platform specific defines

**Comment by hakehuang** [comment]
@NeilChen93 
board testing find below case need to add support.
```
scripts/twister -p frdm_mcxn236 -T /tests/drivers/uart/uart_async_api
```
the board overlay need added.

File: tests/drivers/uart/uart_async_api/boards/frdm_mcxn236.overlay [hunk 14]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright 2024 NXP
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * To test this sample connect P4_0 to P4_1
```
[the remark this answers is not in the document]
Comment (NeilChen93) [reply tests/drivers/uart/uart_async_api/boards/frdm_mcxn236.overlay:7]: I short J5-5(P4_1) and J5-6(P4_0) and test it, this example can work fine. Please check.

[the remark this answers is not in the document]
Comment (hakehuang) [reply tests/drivers/uart/uart_async_api/boards/frdm_mcxn236.overlay:7]: > I short J5-5(P4_1) and J5-6(P4_0) and test it, this example can work fine. Please check.

it is forbidden by electric design, uart is a push/pull I/O and I2C is OC, you should not combine them. and in board design the you should use the J1 2-4.

File: tests/drivers/i2c/i2c_target_api/boards/frdm_mcxn236.overlay [hunk 15]
```diff
@@ -0,0 +1,53 @@
+/*
```
Comment (hakehuang) [inline tests/drivers/i2c/i2c_target_api/boards/frdm_mcxn236.overlay:1]: need change: testcase.yaml adding this platform to:
`drivers.i2c.target_api.single_role:`

File: tests/drivers/uart/uart_async_api/boards/frdm_mcxn236.overlay [hunk 16]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright 2024 NXP
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * To test this sample connect P4_2 to P4_3
+ */
+dut: &flexcomm2_lpuart2 {};
```
Comment (hakehuang) [inline tests/drivers/uart/uart_async_api/boards/frdm_mcxn236.overlay:9]: please do not use lpuart2, as there is a arduino uart interface defined for P4_3 P4_2 which is also can be used for FC2, and this is the board design purpose

**Comment by hakehuang** [comment]
@NeilChen93 , there is a kernel memory testing failure which requires some porting
```
*** Booting Zephyr OS build v3.7.0-rc2-240-gd4616ce20ccd ***
Running TESTSUITE arm_cus**E: ***** BUS FAULT *****
E:   Instruction bus error
E: r0/a1:  0x004000aa  r1/a2:  0x40000000  r2/a3:  0x0000000f
E: r3/a4:  0x000002ec r12/ip:  0x30001688 r14/lr:  0xfffffffd
E:  xpsr:  0x09000049
E: Faulting instruction address (r15/pc): 0x71e3eb50
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Fault during interrupt handling
E: Current thread: 0x30000298 (main)
E: Halting system

```

**Comment by hakehuang** [comment]
|tests/arch/arm/arm_sw_vector_relay/arch.arm.sw_vector_relay | v3.7.0-rc2-242-g17da651f966c | failed | Timeout | frdm_mcxn236 | tests/arch/arm/arm_sw_vector_relay
-- | -- | -- | -- | -- | --

**Comment by NeilChen93** [comment]
Hi @hakehuang , I fixed the issue you mentioned, May I ask where I can see the issue you reported? That way I can check it myself. Thanks.

**Comment by hakehuang** [comment]
> Hi @hakehuang , I fixed the issue you mentioned, May I ask where I can see the issue you reported? That way I can check it myself. Thanks.

let me trigger a new round of testing

**Comment by dleach02** [comment]
@NeilChen93 I forced pushed a fix to the merge conflict on west.yml

File: soc/nxp/mcx/mcxn23x/Kconfig.defconfig
Comment (decsny) [inline soc/nxp/mcx/mcxn23x/Kconfig.defconfig:1]: this file is exactly the same as the one in the mcxnx4x folder

File: soc/nxp/mcx/mcxn23x/Kconfig
Comment (decsny) [inline soc/nxp/mcx/mcxn23x/Kconfig:1]: this file has basically the same code as the mcxn4x folder just with new kconfigs

File: soc/nxp/mcx/mcxn23x/Kconfig.soc
Comment (decsny) [inline soc/nxp/mcx/mcxn23x/Kconfig.soc:1]: why defining a new series?

File: soc/nxp/mcx/mcxn23x/pinctrl_soc.h
Comment (decsny) [inline soc/nxp/mcx/mcxn23x/pinctrl_soc.h:1]: this file looks exactly the same as the one in the mcxnx4x folder

File: soc/nxp/mcx/mcxn23x/soc.c
Comment (decsny) [inline soc/nxp/mcx/mcxn23x/soc.c:1]: this is also another file that is exactly the same as the one in the mcxnx4x folder

File: soc/nxp/mcx/mcxn23x/soc.h
Comment (decsny) [inline soc/nxp/mcx/mcxn23x/soc.h:1]: everything in here is in the mcxnx4x file

File: soc/nxp/mcx/soc.yml [hunk 17]
```diff
@@ -7,6 +7,9 @@ family:
       cpuclusters:
       - name: cpu0
       - name: cpu1
+  - name: mcxn23x
```
Comment (decsny) [inline soc/nxp/mcx/soc.yml:10]: why are we making a new series for every SOC? can we please change this so that the series is just "mcxn" and put both SOC under that series, instead of copy pasting code all over the place that will require duplicated maintenance effort

File: soc/nxp/mcx/mcxn23x/CMakeLists.txt
Comment (decsny) [inline soc/nxp/mcx/mcxn23x/CMakeLists.txt:1]: looks like everything in this file is in the mcxnx4x cmake list as well

**Review by decsny — CHANGES_REQUESTED** [review]
please, let's change the mcx series to "mcxn" and put the SOCs under the same series instead of copy pasting every file for every SOC

**Comment by dleach02** [comment]
> please, let's change the mcx series to "mcxn" and put the SOCs under the same series instead of copy pasting every file for every SOC

@decsny would it be better to have a MCX top-level dir with MCXN/MCXA/MCXL/MCXW as subdirectories like we do with the imxrt directory?

**Comment by NeilChen93** [comment]
Hi @decsny , I update the mcx series to "mcxn" according to your comment, please help double check again, thanks a lot.

File: dts/arm/nxp/nxp_mcxn23x_common.dtsi
Comment (decsny) [inline dts/arm/nxp/nxp_mcxn23x_common.dtsi:1]: similar comments now about the devicetree, is it possible to extract a lot of this into a "common" mcxn file, similar to how there is like an RT10xx.dtsi for the RT series? maybe by renaming the other one and if needed, deleting (or adding) nodes for the mcx236?

File: drivers/timer/mcux_os_timer.c (hunk 12 above)
Comment (decsny) [reply drivers/timer/mcux_os_timer.c:19]: it looks like it's because MCXN236 has no fsl_power.h, for some reason, I'm not sure why though

File: soc/nxp/mcx/mcxn/Kconfig.soc [hunk 18]
```diff
@@ -5,8 +5,12 @@ config SOC_SERIES_MCXNX4X
 	bool
 	select SOC_FAMILY_NXP_MCX
 
+config SOC_SERIES_MCXN23X
```
Comment (decsny) [inline soc/nxp/mcx/mcxn/Kconfig.soc:8]: change also the Kconfigs so that it is SOC_SERIES_MCXN and used by both please

File: drivers/watchdog/wdt_mcux_wwdt.c [hunk 19]
```diff
@@ -83,7 +83,7 @@ static int mcux_wwdt_install_timeout(const struct device *dev,
 	}
 
 #if defined(CONFIG_SOC_MIMXRT685S_CM33) || defined(CONFIG_SOC_MIMXRT595S_CM33) \
-	|| defined(CONFIG_SOC_SERIES_MCXNX4X)
+	|| defined(CONFIG_SOC_SERIES_MCXNX4X) || defined(CONFIG_SOC_SERIES_MCXN23X)
```
Comment (decsny) [inline drivers/watchdog/wdt_mcux_wwdt.c:86]: when you change the Kconfig, you'll have to put this change in the soc commit to retain bisectibility (and change any other occurrences)

File: dts/arm/nxp/nxp_mcxn23x_common.dtsi
Comment (NeilChen93) [reply dts/arm/nxp/nxp_mcxn23x_common.dtsi:1]: Hi @decsny , I don't recommend doing this, because mcxn23x and mcxnx4x are not same project. We will have other mcxn projects in the future, We want some separation between the different projects just like LPC55xx series.

Comment (decsny) [reply dts/arm/nxp/nxp_mcxn23x_common.dtsi:1]: okay, I am less concerned about sharing DT anyways, since some people have some vague ambition to someday auto generate SOC DT which would result in no sharing anyways

**Review by decsny — DISMISSED** [review]

**Review by hakehuang — DISMISSED** [review]

File: west.yml [hunk 20]
```diff
@@ -198,7 +198,7 @@ manifest:
       groups:
         - hal
     - name: hal_nxp
-      revision: 150b98fb2632d2660c8eedb5f992bcc72661fdc1
+      revision: pull/415/head
```
Comment (mmahadevan108) [inline west.yml:201]: You can delete this change.

Comment (dleach02) [reply west.yml:201]: Not yet. The Zephyr side PR still needs to merge. When that occurs I will update his PR (and a couple of others)

**Review by dleach02 — APPROVED** [review]

**Review by decsny — APPROVED** [review]
